File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.py
from collections import deque
N=int(input())
home=[list(map(int,input())) for _ in range(N)]

dx=[0,0,-1,1]
dy=[-1,1,0,0]

ans=[] #오름차순으로 단지의 집 개수를 담을 배열

def bfs(x,y):
    q=deque([(x,y)])
    home[x][y]=0
    cnt=1
    while q:
        x,y=q.popleft()
        for i in range(4):
            nx=x+dx[i]
            ny=y+dy[i]
            if 0<=nx<N and 0<=ny<N and home[nx][ny]==1:
                q.append((nx,ny))
                home[nx][ny]=0
                cnt+=1
    return cnt

for i in range(N):
    for j in range(N):
        if home[i][j]==1:
            ans.append(bfs(i,j))

print(len(ans))
ans.sort()
for a in ans:
    print(a)

File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {

    static class Point {
        int r, c;

        public Point(int r, int c) {
            this.r = r;
            this.c = c;
        }
    }

    static int N;
    static int[][] map;
    static boolean[][] visited;
    static int[] dr = {-1, 1, 0, 0};
    static int[] dc = {0, 0, -1, 1};
    static List<Integer> countList;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        N = Integer.parseInt(br.readLine());

        map = new int[N][N];
        visited = new boolean[N][N];

        for (int i = 0; i < N; i++) {
            String input = br.readLine();
            for (int j = 0; j < input.length(); j++) {
                map[i][j] = input.charAt(j) - '0';
            }
        }

        countList = new ArrayList<>();

        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++) {
                if (!visited[i][j] && map[i][j] == 1) {
                    countList.add(bfs(i, j));
                }
            }
        }

        Collections.sort(countList);

        System.out.println(countList.size());
        for (int number : countList) {
            System.out.println(number);
        }
    }

    private static int bfs(int r, int c) {
        Queue<Point> q = new ArrayDeque<>();
        q.offer(new Point(r, c));
        visited[r][c] = true;

        int count = 1;

        while (!q.isEmpty()) {
            Point p = q.poll();

            int cur_r = p.r;
            int cur_c = p.c;

            for (int i = 0; i < 4; i++) {
                int nr = cur_r + dr[i];
                int nc = cur_c + dc[i];

                if (0 <= nr && nr < N && 0 <= nc && nc < N && !visited[nr][nc] && map[nr][nc] == 1) {
                    count++;
                    q.offer(new Point(nr, nc));
                    visited[nr][nc] = true;
                }
            }
        }

        return count;
    }
}